In [138]:
import numpy as np

def coin_toss(p):     
    result = np.random.binomial(1,p) 
    if result == 0:
        experiment = 'T'
    else :
        experiment = 'H'
    return experiment


def position_experiments(): 
    p=0.5
    n=1000       # number of tosses in one experiment
    full_experiments = []

    for i in range(0, n):    
        full_experiments.append(coin_toss(p))
        i+=1
    
    #print(full_experiments)
    
    htt, hth = 0,0
    
    for i in range(0, len(full_experiments)-2):
        if full_experiments[i] == 'H' and full_experiments[i+1] == 'T' and full_experiments[i+2] == 'T' and htt == 0:
            htt = i+1
        elif full_experiments[i] == 'H' and full_experiments[i+1] == 'T' and full_experiments[i+2] == 'H' and hth == 0:
            hth = i+1
        
        if htt != 0 and hth != 0 :
            break
        i+=1
    
    #print('HTT' + str(htt))
    #print('HTH' + str(hth))
    return [htt, hth]


def Average(lst): 
    return sum(lst) / len(lst)

In [139]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

print('The list of numbers of tosses until HTT :')
print(HTT_position)
print()
print('The list of numbers of tosses until HTH :')
print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print()
print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The list of numbers of tosses until HTT :
[10, 30, 5, 7, 1, 5, 20, 14, 13, 4, 4, 7, 2, 1, 14, 9, 7, 5, 6, 3, 3, 5, 2, 1, 1, 7, 2, 10, 1, 1, 5, 6, 1, 18, 6, 12, 1, 19, 2, 7, 7, 8, 3, 5, 3, 5, 6, 20, 21, 3, 1, 4, 3, 1, 11, 4, 11, 3, 9, 5, 5, 6, 5, 3, 13, 2, 6, 16, 3, 5, 10, 7, 10, 1, 7, 10, 9, 1, 5, 3, 2, 5, 21, 4, 8, 2, 9, 2, 2, 3, 13, 10, 2, 1, 8, 3, 2, 4, 8, 3, 5, 3, 6, 8, 4, 6, 5, 4, 10, 3, 2, 4, 7, 3, 5, 1, 2, 3, 6, 7, 8, 1, 1, 9, 2, 2, 5, 4, 4, 3, 3, 8, 14, 4, 6, 5, 8, 3, 5, 4, 2, 13, 2, 6, 1, 4, 6, 3, 7, 2, 14, 1, 2, 18, 3, 2, 5, 5, 3, 10, 9, 2, 1, 6, 4, 9, 4, 4, 4, 4, 1, 4, 21, 12, 6, 4, 3, 6, 3, 2, 4, 9, 2, 3, 11, 7, 9, 12, 2, 6, 1, 3, 5, 2, 17, 13, 2, 4, 3, 5, 4, 4, 6, 1, 2, 8, 1, 7, 3, 8, 4, 6, 1, 2, 1, 1, 15, 6, 5, 16, 12, 9, 15, 11, 3, 1, 3, 17, 5, 3, 1, 11, 3, 1, 5, 1, 1, 4, 6, 5, 6, 14, 4, 3, 5, 5, 3, 3, 9, 5, 1, 6, 5, 3, 7, 5, 8, 12, 8, 4, 5, 12, 1, 5, 21, 2, 1, 11, 3, 18, 19, 7, 7, 3, 1, 10, 3, 1, 2, 2, 4, 2, 10, 9, 6, 12, 13, 2, 3, 7, 7, 1, 4, 13, 8, 3, 4, 7, 12, 1, 17,

In [140]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 6.053
The average number of tosses until HTH : 8.226


In [141]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 5.906
The average number of tosses until HTH : 8.431


In [142]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 6.196
The average number of tosses until HTH : 7.819


In [143]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 5.976
The average number of tosses until HTH : 7.786


In [144]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 5.828
The average number of tosses until HTH : 8.349


In [145]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 5.909
The average number of tosses until HTH : 8.454


# Report

For this problem, I run 1000 experiments with 1000 tosses each. Moreover, I also run this code 7 times. The result is as follows :

First run :
<br>
The average number of tosses until HTT : 5.962
<br>
The average number of tosses until HTH : 7.335

Second run :
<br>
The average number of tosses until HTT : 6.053
<br>
The average number of tosses until HTH : 8.226

Third run :
<br>
The average number of tosses until HTT : 5.906
<br>
The average number of tosses until HTH : 8.431

Fourth run :
<br>
The average number of tosses until HTT : 6.196
<br>
The average number of tosses until HTH : 7.819

Fith run :
<br>
The average number of tosses until HTT : 5.976
<br>
The average number of tosses until HTH : 7.786

Sixth run :
<br>
The average number of tosses until HTT : 5.828
<br>
The average number of tosses until HTH : 8.349

Seventh run :
<br>
The average number of tosses until HTT : 5.909
<br>
The average number of tosses until HTH : 8.454

## Conclusion
It can be seen that the average number of tosses until HTH is larger than the average number of tosses until HTT, especially when we have a large number of experiments (converge).